File: pkcrypt.py
#!/usr/bin/env python3

import sys
from Crypto.Cipher import AES # requires pycrypto

if len(sys.argv) < 5:
    print("Usage: pkcrypt.py sym_encrypt|sym_decrypt <key> <in> <out>")
    exit(1)

task = sys.argv[1]
if task not in ("sym_encrypt", "sym_decrypt"):
    print("Error: First argument must be sym_encrypt or sym_decrypt.")
    exit(1)

key_filename = sys.argv[2]
in_filename = sys.argv[3]
out_filename = sys.argv[4]

with open(key_filename, "rb") as f:
    pem_data = f.read()
with open(in_filename, "rb") as f:
    data_in = f.read()

# Going for the popular choice...
IV = b"\x00" * AES.block_size

# Just take a random chunk out of the file and use it as our key
key = pem_data[0x20:0x30]
cipher = AES.new(key, AES.MODE_CBC, IV)

if task == "sym_encrypt":
    padding_length = AES.block_size - (len(data_in) % AES.block_size)
    if padding_length != AES.block_size:
        padding_byte = padding_length.to_bytes(1, "big")
        data_in += padding_byte * padding_length
    data_out = cipher.encrypt(data_in)
elif task == "sym_decrypt":
    data_out = cipher.decrypt(data_in)
    # Padding is a badly implemented PKCS#7 where 16 bytes padding is ignored,
    # so we have to check all previous bytes to see if it is valid.
    padding_length = data_out[-1]
    if (padding_length < AES.block_size) && (padding_length < len(data_out)):
        for i in range(0, padding_length):
            if data_out[-1 - i] != padding_length:
                break
        else:
            data_out = data_out[:-padding_length]
else:
    raise NotImplementedError("Operation not supported")

with open(out_filename, "wb") as f:
    f.write(data_out)

File: pkcrypt.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <cstddef>

// Link with cryptopp
#include "cryptopp/aes.h"
#include "cryptopp/modes.h"


const int BLOCKSIZE = CryptoPP::AES::BLOCKSIZE;
// All zero IV
const byte IV[BLOCKSIZE] = {0x00};

int read_file(const std::string& filename, std::vector<char>& data)
{
    std::ifstream stream(filename, std::ios::in | std::ios::binary | std::ios::ate);
    if (!stream.is_open()) {
        return 1;
    }
    std::size_t file_size(stream.tellg());
    stream.seekg(0);
    data.resize(file_size);
    stream.read(data.data(), file_size);
    return 0;
}

int write_file(const std::string& filename, const std::vector<char>& data)
{
    std::ofstream stream(filename, std::ios::out | std::ios::binary);
    if (!stream.is_open()) {
        return 1;
    }
    stream.write(data.data(), data.size());
    return 0;
}

int main(int argc, char **argv)
{
    if (argc < 5) {
        std::cerr << "Usage: pkcrypt sym_encrypt|sym_decrypt <key> <in> <out>\n";
        return 1;
    }

    std::string mode(argv[1]);
    if ((mode != "sym_encrypt") && (mode != "sym_decrypt")) {
        std::cerr << "Error: First argument must be sym_encrypt or sym_decrypt.\n";
        return 1;
    }

    const char* key_filename(argv[2]);
    const char* in_filename(argv[3]);
    const char* out_filename(argv[4]);

    std::vector<char> pem_data;
    if (read_file(key_filename, pem_data)) {
        std::cerr << "Error: Failed to read key file\n";
        return 1;
    }
    if (pem_data.empty()) {
        std::cerr << "Error: PEM file is empty\n";
        return 1;
    }

    std::vector<char> in_data;
    if (read_file(in_filename, in_data)) {
        std::cerr << "Error: Failed to read input file\n";
        return 1;
    }
    if (in_data.empty()) {
        std::cerr << "Error: Input file is empty\n";
        return 1;
    }

    std::vector<char> out_data(in_data.size());

    const char* key = pem_data.data() + 32;

    if (mode == "sym_encrypt") {
        char padding_length = BLOCKSIZE - (in_data.size() % BLOCKSIZE);
        if (padding_length != BLOCKSIZE) {
            in_data.resize(in_data.size() + padding_length, padding_length);
        }
        CryptoPP::CBC_Mode<CryptoPP::AES>::Encryption cipher;
        cipher.SetKeyWithIV((const byte*)key, 16, IV);
        cipher.ProcessData(
            (byte*)out_data.data(), (const byte*)in_data.data(), in_data.size()
        );
    } else {
        CryptoPP::CBC_Mode<CryptoPP::AES>::Decryption cipher;
        cipher.SetKeyWithIV((const byte*)key, 16, IV);
        cipher.ProcessData(
            (byte*)out_data.data(), (const byte*)in_data.data(), in_data.size()
        );
        char padding_length = out_data.back();
        if ((padding_length < BLOCKSIZE) && (padding_length < out_data.size())) {
            bool padding_valid = true;
            for (int i = 0; i < padding_length; i++) {
                if (out_data[i] != padding_length) {
                    padding_valid = false;
                    break;
                }
            }
            if (padding_valid) {
                out_data.resize(out_data.size() - padding_length);
            }
        }
    }

    if (write_file(out_filename, out_data)) {
        std::cerr << "Error: Failed to write output file\n";
        return 1;
    }
    return 0;
}
